File: src/http/config_validation_test.go
package http

import (
	"encoding/json"
	"testing"

	"smartping/src/g"
	"smartping/src/static"
)

func validTestConfig() g.Config {
	return g.Config{
		Port: 8899,
		Name: "local",
		Addr: "127.0.0.1",
		Mode: map[string]string{"Type": "local"},
		Base: map[string]int{
			"Timeout": 5,
			"Refresh": 1,
			"Archive": 30,
		},
		Topology: map[string]string{
			"Tline":       "1",
			"Tsymbolsize": "70",
		},
		Network: map[string]g.NetworkMember{
			"127.0.0.1": {Name: "local", Addr: "127.0.0.1"},
		},
	}
}

func TestValidateConfigAcceptsValidConfig(t *testing.T) {
	if err := validateConfig(validTestConfig()); err != nil {
		t.Fatalf("validateConfig returned error: %v", err)
	}
}

func TestValidateConfigAcceptsEmbeddedDefaultConfig(t *testing.T) {
	data, err := static.Files.ReadFile("conf/config-base.json")
	if err != nil {
		t.Fatalf("read embedded config: %v", err)
	}
	var config g.Config
	if err := json.Unmarshal(data, &config); err != nil {
		t.Fatalf("decode embedded config: %v", err)
	}
	if err := validateConfig(config); err != nil {
		t.Fatalf("embedded default config should be valid: %v", err)
	}
}

func TestValidateConfigRejectsInvalidReferencesAndLimits(t *testing.T) {
	tests := map[string]func(*g.Config){
		"non numeric topology size": func(config *g.Config) {
			config.Topology["Tsymbolsize"] = "large"
		},
		"non finite topology size": func(config *g.Config) {
			config.Topology["Tsymbolsize"] = "NaN"
		},
		"address with whitespace": func(config *g.Config) {
			config.Addr = " 127.0.0.1"
		},
		"network key mismatch": func(config *g.Config) {
			config.Network["127.0.0.1"] = g.NetworkMember{Name: "local", Addr: "127.0.0.2"}
		},
		"missing self node": func(config *g.Config) {
			config.Addr = "127.0.0.2"
		},
		"missing ping target": func(config *g.Config) {
			member := config.Network["127.0.0.1"]
			member.Ping = []string{"192.0.2.1"}
			config.Network["127.0.0.1"] = member
		},
		"invalid auth IP": func(config *g.Config) {
			config.Authiplist = "127.0.0.1,invalid"
		},
		"excessive ping count": func(config *g.Config) {
			config.Base["PingCount"] = 1000000
		},
		"invalid cloud endpoint": func(config *g.Config) {
			config.Mode = map[string]string{"Type": "cloud", "Endpoint": "file:///etc/passwd"}
		},
		"zero port": func(config *g.Config) {
			config.Port = 0
		},
		"port above maximum": func(config *g.Config) {
			config.Port = 65536
		},
	}

	for name, mutate := range tests {
		t.Run(name, func(t *testing.T) {
			config := validTestConfig()
			mutate(&config)
			if err := validateConfig(config); err == nil {
				t.Fatalf("validateConfig should reject %s", name)
			}
		})
	}
}

func TestValidateConfigAcceptsValidTopologyRule(t *testing.T) {
	config := validTestConfig()
	config.Network["192.0.2.1"] = g.NetworkMember{Name: "remote", Addr: "192.0.2.1"}
	member := config.Network["127.0.0.1"]
	member.Ping = []string{"192.0.2.1"}
	member.Topology = []map[string]string{{
		"Name":        "remote",
		"Addr":        "192.0.2.1",
		"Thdchecksec": "900",
		"Thdloss":     "30",
		"Thdavgdelay": "200",
		"Thdoccnum":   "3",
	}}
	config.Network["127.0.0.1"] = member

	if err := validateConfig(config); err != nil {
		t.Fatalf("validateConfig returned error: %v", err)
	}
}

func TestValidateConfigRejectsNonMinuteTopologyWindow(t *testing.T) {
	config := validTestConfig()
	config.Network["192.0.2.1"] = g.NetworkMember{Name: "remote", Addr: "192.0.2.1"}
	member := config.Network["127.0.0.1"]
	member.Topology = []map[string]string{{
		"Name":        "remote",
		"Addr":        "192.0.2.1",
		"Thdchecksec": "61",
		"Thdloss":     "30",
		"Thdavgdelay": "200",
		"Thdoccnum":   "3",
	}}
	config.Network["127.0.0.1"] = member

	if err := validateConfig(config); err == nil {
		t.Fatalf("validateConfig should reject a topology window that is not a multiple of 60 seconds")
	}
}

File: web/src/utils/configValidation.test.ts
import assert from 'node:assert/strict'
import test from 'node:test'
import type { Config } from '../types/index.js'
import { CONFIG_LIMITS, isValidIPv4, validateConfigForEdit } from './configValidation.js'

const makeConfig = (): Config => ({
  Ver: 'test',
  Port: 8899,
  Name: 'Local',
  Addr: '192.0.2.1',
  Mode: { Type: 'local', Endpoint: '' },
  Base: {
    Timeout: 5,
    Archive: 10,
    Refresh: 1,
    PingCount: 20,
    PingIntervalMs: 3000,
    PingTimeoutMs: 3000,
    PingStaggerMs: 100,
    MappingConcurrency: 8,
    MappingProbeCount: 3
  },
  Topology: { Tline: '1', Tsymbolsize: '70', Tsound: '/alert.mp3' },
  Network: {
    '192.0.2.1': {
      Name: 'Local',
      Addr: '192.0.2.1',
      Smartping: true,
      Ping: ['198.51.100.2'],
      Topology: [
        {
          Name: 'Remote',
          Addr: '198.51.100.2',
          Thdchecksec: '900',
          Thdoccnum: '3',
          Thdavgdelay: '200',
          Thdloss: '30'
        }
      ]
    },
    '198.51.100.2': {
      Name: 'Remote',
      Addr: '198.51.100.2',
      Smartping: true,
      Ping: [],
      Topology: []
    }
  },
  Chinamap: {
    Shanghai: {
      ctcc: ['203.0.113.1'],
      cucc: [],
      cmcc: []
    }
  },
  Toollimit: 0,
  Authiplist: '127.0.0.1,2001:db8::1'
})

test('validateConfigForEdit accepts a complete valid configuration', () => {
  assert.equal(validateConfigForEdit(makeConfig()), null)
})

test('isValidIPv4 rejects ambiguous and malformed addresses', () => {
  for (const address of ['127.0.0.1', '0.0.0.0', '255.255.255.255']) {
    assert.equal(isValidIPv4(address), true, address)
  }
  for (const address of ['127.0.0.01', '256.0.0.1', '127.0.0', ' 127.0.0.1', '::1']) {
    assert.equal(isValidIPv4(address), false, address)
  }
})

test('validateConfigForEdit rejects invalid optional base values', () => {
  const config = makeConfig()
  config.Base.MappingConcurrency = Number.NaN

  assert.equal(validateConfigForEdit(config)?.key, 'config.validationBaseParameter')
})

test('topology display controls do not offer values rejected by validation', () => {
  assert.equal(CONFIG_LIMITS.topologyLine.min > 0, true)
  assert.equal(CONFIG_LIMITS.topologySymbol.min > 0, true)

  const zeroLine = makeConfig()
  zeroLine.Topology.Tline = '0'
  assert.equal(validateConfigForEdit(zeroLine)?.key, 'config.validationLineWidth')

  const zeroSymbol = makeConfig()
  zeroSymbol.Topology.Tsymbolsize = '0'
  assert.equal(validateConfigForEdit(zeroSymbol)?.key, 'config.validationSymbolSize')
})

test('validateConfigForEdit rejects non-decimal topology values', () => {
  for (const value of ['0x10', '0b10', '0o10', '0x1p4']) {
    const config = makeConfig()
    config.Topology.Tline = value
    assert.equal(validateConfigForEdit(config)?.key, 'config.validationLineWidth', value)
  }

  for (const value of ['.5', '1.', '+1', '1e1']) {
    const config = makeConfig()
    config.Topology.Tline = value
    assert.equal(validateConfigForEdit(config), null, value)
  }
})

test('validateConfigForEdit rejects duplicate ping targets', () => {
  const config = makeConfig()
  config.Network[config.Addr]!.Ping.push('198.51.100.2')

  assert.equal(validateConfigForEdit(config)?.key, 'config.validationDuplicateTarget')
})

test('validateConfigForEdit enforces topology sample-window constraints', () => {
  const config = makeConfig()
  config.Network[config.Addr]!.Topology[0]!.Thdchecksec = '120'
  config.Network[config.Addr]!.Topology[0]!.Thdoccnum = '3'

  assert.equal(validateConfigForEdit(config)?.key, 'config.validationTopologyRule')
})

test('validateConfigForEdit rejects unsafe cloud endpoints', () => {
  const endpoints = [
    'file:///tmp/config.json',
    'https://user:secret@example.com/config.json',
    'https://@example.com/config.json',
    'https://:@example.com/config.json',
    'https://example.com/config.json#private',
    'https://example.com:65536/config.json'
  ]

  for (const endpoint of endpoints) {
    const config = makeConfig()
    config.Mode = { Type: 'cloud', Endpoint: endpoint }
    assert.equal(validateConfigForEdit(config)?.key, 'config.validationCloudEndpoint', endpoint)
  }
})

test('validateConfigForEdit rejects unsupported mapping carriers', () => {
  const config = makeConfig()
  const providers = config.Chinamap.Shanghai as unknown as Record<string, string[]>
  providers.unknown = ['203.0.113.2']

  assert.equal(validateConfigForEdit(config)?.key, 'config.validationMappingProvider')
})
